plugins/multifundchannel.c

```diff
@@ -0,0 +1,1643 @@
+	be announced.
+	*/
+	bool announce;
+	/* How much of the initial funding to push to
+	the destination.
+	*/
+	struct amount_msat push_msat;
+
+	/* The actual channel_id.  */
+	const char *channel_id;
+
+	/* The spark currently running for this destination.  */
+	struct plugin_spark_completion *spark;
+
+	/* Any error messages.  */
+	const char *error;
+};
+
+/* The object for a single multifundchannel command.  */
+struct multifundchannel_command {
+	/* The plugin-level command.  */
+	struct command *cmd;
+	/* An array of destinations.  */
+	struct multifundchannel_destination *destinations;
+	/* An array of sparks running each destination.  */
+	struct plugin_spark **sparks;
+
+	/* The feerate desired by the user.  */
+	const char *feerate_str;
+	/* The minimum number of confirmations for owned
+	UTXOs to be selected.
+	*/
+	u32 minconf;
+	/* The set of utxos to be used.  */
+	const char *utxos_str;
+
+	/* Flag set when any of the destinations has a value of "all".  */
+	bool has_all;
+
+	/* The txid of the funding transaction.
+	This can be either the "dry run" transaction which
+	reserves the funds, or the final funding transaction.
+	*/
+	struct bitcoin_txid *txid;
+
+	/* The actual tx of the actual final funding transaction
+	that was broadcast.
+	*/
+	const char *final_tx;
+	const char *final_txid;
+};
+
+extern const struct chainparams *chainparams;
+
+/*-----------------------------------------------------------------------------
+Command Cleanup
+-----------------------------------------------------------------------------*/
+
+/*~
+We disallow the use of command_fail and forward_error directly
+in the rest of the code.
+
+This ensures that if we ever fail a multifundchannel, we do cleanup
+by doing fundchannel_cancel and txdiscard.
+*/
+
+/* TODO: This is lengthy enough to deserve its own source file,
+clocking in at 240 loc.
+*/
+
+/* Object for performing cleanup.  */
+struct multifundchannel_cleanup {
+	struct plugin_spark **sparks;
+	struct command_result *(*cb)(void *arg);
+	void *arg;
+};
+
+/* Spark function that cleans up a `txprepare`d txid.  */
+static struct command_result *
+mfc_cleanup_txid_spark(struct command *cmd,
+		       struct plugin_spark_completion *comp,
+		       struct bitcoin_txid *txid);
+/* Spark function that cleans up a `fundchannel_start`ed node id.  */
+static struct command_result *
+mfc_cleanup_fc_spark(struct command *cmd,
+		     struct plugin_spark_completion *comp,
+		     const char *nodeid);
+/* Run at completion of all cleanup sparks.  */
+static struct command_result *
+mfc_cleanup_complete(struct command *cmd,
+		     struct multifundchannel_cleanup *cleanup);
+
+/* Core cleanup function.  */
+static struct command_result *
+mfc_cleanup_(struct multifundchannel_command *mfc,
+	     struct command_result *(*cb)(void *arg),
+	     void *arg)
+{
+	struct multifundchannel_cleanup *cleanup;
+	unsigned int i;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %p: cleanup!", mfc);
```

Comment:
It is useful in case somebody is running multiple such commands in parallel; that way we have a unique identifier to separate which debug log goes to which command.  Of course, given that malloc implementation will reuse memory, it is not so perfectly unique; perhaps an autoincrement id would do better.

Reply:
OR getting the `struct command::id`; this is unique for the plugin, right?  That might be a better tag for knowing *which* `multifundchannel` we are talking about on the logs.